### DISCUSSION TITLE
Function 'extract_feature_edges' failed to extract all the feature edges.

### DISCUSSION THREAD

**Discussion by pic-wang** [pyvista/pyvista#6747] [Q&A] [answered]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [x] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [X] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

```python
import pyvista as pv
import numpy as np
import sys

stl_filename = "../constant/triSurface/buildings_wrong.stl"
mesh = pv.read(stl_filename)

# Extract point and face from stl
points = mesh.points
print("Point shape:",points.shape)
face_point_idx_before = mesh.faces # stl file contains triangles, so each face has 3 points
print(face_point_idx_before.shape)

face_point_idx_1 = mesh.faces.reshape(-1, 4)
face_point_idx = mesh.faces.reshape(-1, 4)[:, 1:]

print("First element before slicing", face_point_idx_1[0])
print(face_point_idx_1.shape)
print("First element after slicing", face_point_idx[0])
print("Total face:", face_point_idx.shape) # Face number: 152


# Get face at bottom

z0_point_indices = np.where(np.isclose(points[:, 2], 0.0))[0] # np.where returns a tuple, we only need the first element
print("z0_point_indices", z0_point_indices.shape)

z0_point_set = set(z0_point_indices) # Use set to make the search faster
print("z0_point_set", len(z0_point_set))
z0_faces_indices = []
for i, face in enumerate(face_point_idx):
    if all(idx in z0_point_set for idx in face):
        z0_faces_indices.append(i)
print("Shape of z0_faces_indices", len(z0_faces_indices))
print("z0_faces_indices", z0_faces_indices)


z0_mesh = mesh.extract_cells(z0_faces_indices) # Return a new mesh with only the selected cells
print("z0_mesh", z0_mesh)
edges = z0_mesh.extract_feature_edges()
edges_all = mesh.extract_feature_edges()

# # Plot the features edges of ground will miss one edges
# p = pv.Plotter()
# p.add_mesh(z0_mesh, color=True)
# p.add_mesh(edges, color="red", line_width=5)
# p.camera.zoom(1.5)
# p.show()

# Plot all the features edges works
p = pv.Plotter()
p.add_mesh(mesh, color=True)
p.add_mesh(edges_all, color="red", line_width=5)
p.camera.zoom(1.5)
p.show()
```

### Description

I'm using PyVista to extract the feature edges of the ground face of a building model. The code successfully extracts the ground face, but the resulting feature edges are incomplete (some edges are missing). However, when I extract the feature edges for the entire building model, the edges are correctly captured in full. I’d like to achieve complete feature edge extraction specifically for the ground face. I also attach my geomety here. Any reply or discussion is helpful!!
[buildings_wrong.zip](https://github.com/user-attachments/files/17555035/buildings_wrong.zip)
![image](https://github.com/user-attachments/assets/84f36963-c580-4c77-8fe4-ae8a30e14317)
![image](https://github.com/user-attachments/assets/b04d5487-c8c1-4dbe-b542-237da8937189)

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Tue Oct 29 18:04:22 2024 中国标准时间

                OS : Windows (10 10.0.19041 SP0 Multiprocessor Free)
            CPU(s) : 16
           Machine : AMD64
      Architecture : 64bit
               RAM : 31.8 GiB
       Environment : Jupyter
        GPU Vendor : NVIDIA Corporation
      GPU Renderer : NVIDIA GeForce RTX 4060 Ti/PCIe/SSE2
       GPU Version : 4.5.0 NVIDIA 560.94
  MathText Support : False

  Python 3.9.19 | packaged by conda-forge | (main, Mar 20 2024, 12:38:46) [MSC
  v.1929 64 bit (AMD64)]

           pyvista : 0.44.1
               vtk : 9.3.1
             numpy : 1.26.3
        matplotlib : 3.9.2
            scooby : 0.10.0
             pooch : 1.8.2
            pillow : 10.2.0
...
      trame_client : 3.4.0
      trame_server : 3.2.3
      nest_asyncio : 1.6.0
--------------------------------------------------------------------------------
```

**Comment by darikg** [comment, marked as answer]
Thanks for attaching the mesh! This was difficult to track down but I believe it's correct behavior.

If we label the vertices we can see that the edge of interest is between vertices 2 and 8:

```python
# Plot the features edges of ground will miss one edges
p = pv.Plotter(window_size=(500, 500))
p.add_mesh(z0_mesh, color=True, show_edges=True, opacity=0.5)
p.add_mesh(edges, color="red", line_width=5)
p.add_point_labels(z0_mesh.points, [str(i) for i in range(z0_mesh.n_points)])
p.show()
```

![image](https://github.com/user-attachments/assets/aaed53fa-049c-4db7-bfbe-4c7f5e16fcd5)

And find that edge belongs to two triangles, (8,1, 2) and (28, 2, 8):

```python
that_edge = (2, 8)
tris = z0_mesh.cells.reshape(-1, 4)[:, 1:]
has_that_edge = np.isin(tris, that_edge).sum(axis=1) == 2
print(tris[has_that_edge])
```
>[[ 8  1  2]
 [28  2  8]]

Because the edge belongs to two triangles, it's not a border edge. And because both triangles traverse the edge in the same direction (from 2 to 8), they have opposite orientation. This means that the dihedral angle between the two faces is 0 rather than 180 degrees, so it's not a feature edge either.

You can convert the unstructured grid `z0_mesh` to a `PolyData` object to use the `compute_normals` function to make sure adjacent triangles have the same orientation:

```python
m = pv.PolyData(z0_mesh.points, faces=z0_mesh.cells)
m = m.compute_normals(consistent_normals=True)
p = pv.Plotter()
p.add_mesh(m)
p.add_mesh(m.extract_feature_edges(), color="red", line_width=5)
p.show()
```
![image](https://github.com/user-attachments/assets/4e6da6f0-9ccc-43d5-bde0-d387d6851756)

**Reply by pic-wang** [reply]
Your reply is amazing! Thanksss!
